File: org.kevoree.modeling.cpp.microframework/src/main/java/org/kevoree/modeling/api/trace/ModelTrace.kt
package org.kevoree.modeling.api.trace

import org.kevoree.modeling.api.util.ActionType


/**
 * Created by duke on 25/07/13.
 */

trait ModelTrace {
    fun toString(): String
}

class ModelAddTrace(val srcPath: String, val refName: String, val previousPath: String?, val typeName: String?): ModelTrace {
    override fun toString(): String {
        val buffer = StringBuilder()
        buffer.append("{ \"traceType\" : " + ActionType.ADD + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\"")
        if(previousPath != null){
            buffer.append(", \"previouspath\" : \"" + previousPath + "\"")
        }
        if(typeName != null){
            buffer.append(", \"typename\" : \"" + typeName + "\"")
        }
        buffer.append("}")
        return buffer.toString()
    }
}

class ModelAddAllTrace(val srcPath: String, val refName: String, val previousPath: List<String>?, val typeName: List<String>?): ModelTrace {

    private fun mkString(ss: List<String>?): String? {
        if(ss == null){
            return null
        }
        val buffer = StringBuilder()
        var isFirst = true
        for(s in ss){
            if(!isFirst){
                buffer.append(",")
            }
            buffer.append(s)
            isFirst = false
        }
        return buffer.toString()
    }

    override fun toString(): String {
        val buffer = StringBuilder()
        buffer.append("{ \"traceType\" : " + ActionType.ADD_ALL + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\"")
        if(previousPath != null){
            buffer.append(", \"previouspath\" : \"" + mkString(previousPath) + "\"")
        }
        if(typeName != null){
            buffer.append(", \"typename\" : \"" + mkString(typeName) + "\"")
        }
        buffer.append("}")
        return buffer.toString()
    }
}

class ModelRemoveTrace(val srcPath: String, val refName: String, val objPath: String): ModelTrace {
    override fun toString(): String {
        return "{ \"traceType\" : " + ActionType.REMOVE + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\", \"objpath\" : \"" + objPath + "\" }"
    }
}

class ModelRemoveAllTrace(val srcPath: String, val refName: String): ModelTrace {
    override fun toString(): String {
        return "{ \"traceType\" : " + ActionType.REMOVE_ALL + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\" }"
    }
}

class ModelSetTrace(val srcPath: String, val refName: String, val objPath: String?, val content: String?, val typeName: String?): ModelTrace {
    override fun toString(): String {
        val buffer = StringBuilder()
        buffer.append("{ \"traceType\" : " + ActionType.SET + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\"")
        if(objPath != null){
            buffer.append(", \"objpath\" : \"" + objPath + "\"")
        }
        if(content != null){
            buffer.append(", \"content\" : \"" + content + "\"")
        }
        if(typeName != null){
            buffer.append(", \"typename\" : \"" + typeName + "\"")
        }
        buffer.append("}")
        return buffer.toString()
    }

}

File: org.kevoree.modeling.cpp.microframework/src/main/cpp/microframework/api/trace/ModelTrace.cpp
#include <microframework/api/trace/ModelTrace.h>

/**
 * Author: jedartois@gmail.com
 * Date: 24/10/13
 * Time: 18:36
 */
ModelSetTrace::ModelSetTrace (string _srcPath, string _refName, string _objPath, string _content, string _typeName)
{
	srcPath = _srcPath;
	refName = _refName;
	typeName = _typeName;
	content = _content;
	typeName = _typeName;
}


ModelSetTrace::~ModelSetTrace(){

}
std::string ModelSetTrace::toString ()
{
	string buffer = "";
	buffer.append ("{ \"traceType\" : " + ActionType (SET) +
			" , \"src\" : \"" + srcPath + "\", \"refname\" : \"" +
			refName + "\"");
	if (!objPath.empty ())
	{
		buffer.append (", \"objpath\" : \"" + objPath + "\"");
	}
	if (!content.empty ())
	{
		buffer.append (", \"content\" : \"" + content + "\"");
	}
	if (!typeName.empty ())
	{
		buffer.append (", \"typename\" : \"" + typeName + "\"");
	}
	buffer.append ("}\n");

	return buffer;

}  


ModelAddTrace::~ModelAddTrace(){

}
ModelAddTrace::ModelAddTrace(string _srcPath, string _refName, string _previousPath,string _typeName)
{
	srcPath  =_srcPath;
	refName=_refName;
	previousPath =_previousPath;
	typeName=_typeName;
}

std::string ModelAddTrace::toString ()
{
	string buffer = "";

	buffer.append ("{ \"traceType\" : " + ActionType (ADD) +
			" , \"src\" : \"" + srcPath + "\", \"refname\" : \"" +
			refName + "\"");
	if (!previousPath.empty ())
	{
		buffer.append (", \"previouspath\" : \"" + previousPath + "\"");
	}
	if (!typeName.empty ())
	{
		buffer.append (", \"typename\" : \"" + typeName + "\"");
	}
	buffer.append ("}\n");
	return buffer;
}


ModelAddAllTrace::~ModelAddAllTrace(){

}

ModelAddAllTrace::ModelAddAllTrace (string _srcPath, string _refName,list < string > &_previousPath, list < string > &_typeName)
{
	srcPath = _srcPath;
	refName = _refName;
	previousPath = _previousPath;
	typeName = _typeName;

}

std::string ModelAddAllTrace::toString ()
{
	string buffer;
	buffer.append ("{ \"traceType\" : " + ActionType (ADD_ALL) +
			" , \"src\" : \"" + srcPath + "\", \"refname\" : \"" +
			refName + "\"");
	if (!previousPath.empty ())
	{
		buffer.append (", \"previouspath\" : \"" + mkString (previousPath) +
				"\"");
	}
	if (typeName.empty ())
	{
		buffer.append (", \"typename\" : \"" + mkString (typeName) + "\"");
	}
	buffer.append ("}\n");
	return buffer;
}

std::string ModelAddAllTrace::mkString (list < string > ss){

	string buffer;
	if (!ss.empty ())
	{
		return buffer;
	}
	std::list < std::string >::iterator it;
	bool isFirst = true;
	for (it = ss.begin (); it != ss.end (); ++it)
	{
		if (!isFirst)
		{
			buffer.append (",");
		}
		buffer.append (*it);
		isFirst = false;
	}
	return buffer;
}


ModelRemoveTrace::ModelRemoveTrace (string _srcPath, string _refName, string _objPath)
{
	srcPath = _srcPath;
	refName = _refName;
	objPath = _objPath;
}


ModelRemoveTrace::~ModelRemoveTrace(){

}
std::string ModelRemoveTrace::toString ()
{
	return "{ \"traceType\" : " + ActionType (REMOVE) + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\", \"objpath\" : \"" + objPath + "\" }\n";
}

ModelRemoveAllTrace::~ModelRemoveAllTrace(){

}


ModelRemoveAllTrace::ModelRemoveAllTrace (string _srcPath, string _refName)
{
	srcPath = _srcPath;
	refName = _refName;
}


std::string ModelRemoveAllTrace::toString ()
{
	return "{ \"traceType\" : " + ActionType (REMOVE_ALL) + " , \"src\" : \"" + srcPath + "\", \"refname\" : \"" + refName + "\" }\n";
}
